6 integrate twilio

chandojo opened the pull request:
Issue #4 Integrate with Twilio API for receiving SMS

[PROBLEM]
Users will send SMS to Twilio phone number which will be consumed by the service.

[SOLUTION]
* Messenger using Twilio client added
* Subscription App created to send correct reponse to subscriber

[TEST]
* Unit tests added for Messenger

Notes:
Trouble squashing commits 1835ff and e5ca936

ErrorsAndGlitches commented on README.md:
```diff
@@ -52,6 +52,9 @@ application should be configured to be a **Regular Web Application**.
 * `AUTH0_SECRET`: The Auth0 application's client secret.
+* `TWILIO_ACCOUNT_SID`: Twilio Account SID for client
+* `TWILIO_AUTH_TOKEN`: Twilio Auth Token for client
+* `TWILIO_SMS_NUMBER`: Phone number bought through Twilio for client
```
* `TWILIO_SMS_NUMBER`: Twilio Phone Number

**bought** is a little weird because you can get these numbers for free for testing.

ErrorsAndGlitches commented on environment-docker.yml:
```diff
@@ -2,57 +2,59 @@ name: waisn-tech-tools
-  - django=2.2.1=py37_0
+  - cffi=1.12.3=py37hb5b8e2f_0
+  - cryptography=2.6.1=py37ha12b0ac_0
+  - django=2.1.4=py37_0
```
I'm seeing a number of packages being downgraded like Django but not in the conda environment configuration for mac. Is this needed?

chandojo replied:
It looks like these were some accidental changes from updating the environment files

ErrorsAndGlitches commented on environment-docker.yml:
```diff
@@ -2,57 +2,59 @@ name: waisn-tech-tools
     - urllib3==1.25.3
-prefix: /opt/conda/envs/waisn-tech-tools
+prefix: //anaconda3/envs/waisn-tech-tools
+
```
This should not change. Conda environments in the Docker image are installed in the **/opt/conda/envs/** location.

ErrorsAndGlitches commented on environment-mac.yml:
```diff
@@ -46,11 +46,15 @@ dependencies:
     - transitions==0.6.9
+    - twilio==6.24.1
     - urllib3==1.25.3
+prefix: //anaconda3/envs/waisn-tech-tools
```
remove - conda environments could be configured differently on different developer's machines hence I usually remove `prefix` in conda package files.

ErrorsAndGlitches replied:
**prefix** is still here.

ErrorsAndGlitches replied:
this is marked as resolved, but the prefix is still here in the latest pull request.

ErrorsAndGlitches commented on waisntechtools/alerts/debug_twilio.py:
```diff
@@ -0,0 +1,24 @@
+import datetime
+
+def debug_twilio():
+    messages=[]
```
I think we can get rid of this based on a latter comment. But wanted to take the opportunity to mention that it is good practice to declare variables near where they are used. Hence, I would move this to line 19. There should be good locality in code.

ErrorsAndGlitches commented on waisntechtools/alerts/debug_twilio.py:
```diff
@@ -0,0 +1,24 @@
+
+        for message in messages_list:
+            messages.append(message)
+        return messages
```
Is there a reason the code can't simply return `messages_list`? I.e.:

```
return client.messages.list(limit=10)
```

ErrorsAndGlitches commented on waisntechtools/alerts/debug_twilio.py:
```diff
@@ -0,0 +1,24 @@
+
+    if settings.DEBUG == True:
+        request_valid = True
+    # TODO: create validator for production settings
```
seems like there is some functionality missing here...

chandojo replied:
Will need to follow up about this.  When we are both free we should go over authorization and security of the project.

Reminder to self:
- Understand 0Auth security process 
- Determine use of Twilio's Request Validator

chandojo replied:
Determined that these particular lines needed to be removed. Twilio documentation recommends RequestValidators when Twilio sends a request **to** the Django app not when receiving an API call.  Credentials are verified through Twilio's on system using our sid and auth token.

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -1,9 +1,11 @@
 from django.shortcuts import render
-from django.views.generic import ListView
+from django.views.generic import ListView, TemplateView
```
Is `ListView` still used?

ErrorsAndGlitches commented on waisntechtools/alerts/subscription_states.py:
```diff
@@ -1,8 +1,9 @@
 from .asset_files import AssetFiles
 
-
 class SubscriptionStates(object):
```
pep 8 style is to have 2 spaces between imports and class. Some editors will point this out for you :)

![](https://i.imgflip.com/3dsamy.jpg)

ErrorsAndGlitches commented on waisntechtools/alerts/subscription_states.py:
```diff
@@ -67,14 +68,14 @@ class SubscriptionStates(object):
 
         self._subscriber = subscriber
         self._machine = Machine(
             model=self,
             states=SubscriptionStates._STATES,
-            initial=subscriber.state,
+            initial=subscriber.subscription_state,
```
thanks for fixing!

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,30 @@
+from alerts.messenger import Messenger, Message
+from alerts.models import Subscriber
+
+# TODO: Create Messenger test to return file name and file contents
```
Should this be removed?

ErrorsAndGlitches commented on waisntechtools/alerts/subscription_app.py:
```diff
@@ -0,0 +1,69 @@
+from .subscription_states import SubscriptionStates
+
+class SubscriptionApp():
+    def handle(request):
```
What do you think about not having a class here i.e. having `def handle()` function by itself? There doesn't seem to be much of a reason to have the class because there are no member variables.

ErrorsAndGlitches commented on waisntechtools/alerts/subscription_app.py:
```diff
@@ -0,0 +1,69 @@
+
+    def receive(self):
+        self._handled_msgs[self._subscriber.subscription_state](self._message_body)
+
```
nit: spacing. Pep 8 is to have one new line between functions in classes.

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,30 @@
+        self.assertEqual(read_message, 'The test has completed')
+
+    def tearDown(self):
+        self.mocked_response = None
```
this isn't needed. Python interpreter will clean up the `mocked_response` variable.

ErrorsAndGlitches commented on waisntechtools/alerts/urls.py:
```diff
@@ -10,4 +11,5 @@
     path('alert/sent', alert_request.sent, name='alert_sent'),
     path('follow-up', follow_up_request.index, name='follow_up'),
     path('follow-up/sent', follow_up_request.sent, name='follow_up_sent'),
+    path('subscribe', csrf_exempt(subscription_app.SubscriptionApp.handle), name='subscribe'),
```
the naming is a little weird, because it might capture more than subscribing. Does it make sense to you to make this a little more generic? Perhaps:

```
path('twilio-endpoint', csrf_exempt(twilio_app.handle), name='twilio_endpoint'),
```

Also, it seems that it might be possible to remove the `csrf_exempt` based on this Twilio article, which seems to be a good idea to do.
* https://www.twilio.com/blog/2014/04/building-a-simple-sms-message-application-with-twilio-and-django-2.html

chandojo replied:
This seems reasonable.  Adjustments have been made in congruence with the article

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,30 @@
+    def test_message(self):
+        with patch('builtins.open', self.mocked_response, create=True) as f:
+            read_message = Message(self.mocked_response).contents()
+        self.assertEqual(read_message, 'The test has completed')
```
the use of `mock_open` and `patch` seems a bit wonky. Do you think it would be a bit more natural by creating a temporary file? It can be configured to delete on interpreter exit.

* [tempfile.NamedTemporaryFile](https://docs.python.org/2/library/tempfile.html#tempfile.NamedTemporaryFile)

chandojo replied:
I tried this tempfile.NamedTemporaryFile first and had issues using it with the Message class.  This was a work around.  It could be refactored in future

ErrorsAndGlitches replied:
Can you share the issue or stack trace?

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,30 @@
+        self.mock_subscriber = Mock()
+        self.mock_subscriber.phone_number = "+11234567890"
+
+    def test_messenger(self):
```
This test doesn't `assert` anything.

chandojo replied:
The initial thought on this test was to test that the messages would be successfully sent without raising error BUT I've removed the test MessengerTestCase because if there is an issue with sending the message, the Twilio dashboard will alert the user with an error

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -25,10 +27,14 @@ def debug(request):
 
-    def get_queryset(self):
-        return Subscriber.objects.order_by('-date_registered')
+    def get_twilio(self, **kwargs):
```
inline?

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -25,10 +27,14 @@ def debug(request):
+
+    def get_context_data(self, **kwargs):
+        context = super(DebugView, self).get_context_data(**kwargs)
+        context['messages'] = self.get_twilio
```
this sets the **'messages'** value to be a function - is that right?

chandojo replied:
Nope, it should be the values returned from the function that is called. Good catch

ErrorsAndGlitches commented on waisntechtools/alerts/tests/fakes.py:
```diff
@@ -52,6 +52,10 @@ def language(self):
             self._rand.randrange(0, len(_LanguageProvider._LANGUAGES))
         ]
 
+class Message(DjangoModelFactory):
```
Is this used?

chandojo replied:
Doesn't look like it

ErrorsAndGlitches commented on waisntechtools/alerts/messenger.py:
```diff
@@ -0,0 +1,28 @@
+
+    def send(self, subscriber, filenames):
+        message_to = subscriber.phone_number
+        for file in filenames:
```
`for filename in filenames`

it's not technically a file.

ErrorsAndGlitches commented on waisntechtools/alerts/messenger.py:
```diff
@@ -0,0 +1,28 @@
+                                      )
+
+class Message(object):
+    def __init__(self, file):
```
`filename`

ErrorsAndGlitches commented on waisntechtools/alerts/debug_twilio.py:
```diff
@@ -0,0 +1,24 @@
+import requests
+import datetime
+
+def debug_twilio():
```
can this method be more specific? Maybe `recent_twilio_messages`?

ErrorsAndGlitches reviewed (commented):
1. Can you squash the commits? It looks like there are three of them in this pull request.
2. Could you update the description in the pull request with any end-2-end tests that you have done? It would be good to verify that the Docker image still builds as well.
3. Could you update the `README` with information about how to set up **ngrok** for local development with Twilio? This is definitely non-trivial and could save me a bunch of time.

ErrorsAndGlitches reviewed (commented):
`ngrok` should not be checked into the repo.

ErrorsAndGlitches commented on README.md:
```diff
@@ -52,6 +52,22 @@ application should be configured to be a **Regular Web Application**.
+## Running NGROK
+NGROK is needed to manually test sending text messages
+
+- In your root directory:
```
while both '-' and '\*' will both create unordered lists, given that the rest of the README uses '*', it would be good to be consistent for readability. I don't personally care which one to use, so if you want to use '-', could you change all the other lists to it?

ErrorsAndGlitches commented on README.md:
```diff
@@ -52,6 +52,22 @@ application should be configured to be a **Regular Web Application**.
+
+- In your root directory:
+
+` ./ngrok http 80 `
```
perhaps a more normal way to write this is without the `./` since it seems to imply that the tool has to be in some directory. Also, I imagine most people would get `ngrok` from a package manager like Home Brew.

ErrorsAndGlitches commented on README.md:
```diff
@@ -52,6 +52,22 @@ application should be configured to be a **Regular Web Application**.
+* `TWILIO_SMS_NUMBER`: Phone number received from Twilio for client
+
+## Running NGROK
+NGROK is needed to manually test sending text messages
```
You should expand on this since `ngrok` by itself doesn't have much to do with sending text messages. `ngrok` is used to create IP tunnels from a host behind a firewall to the public internet, not really much to do with sending and receiving text messages.

I wrote this for my own documentation when I was updating your code. Perhaps you can get some inspiration from it?

* https://github.com/ErrorsAndGlitches/waisn-tech-tools/tree/2-sns-integration#setting-up-twilio

ErrorsAndGlitches commented on README.md:
```diff
@@ -52,6 +52,22 @@ application should be configured to be a **Regular Web Application**.
+* `TWILIO_AUTH_TOKEN`: Twilio Auth Token for client
+* `TWILIO_SMS_NUMBER`: Phone number received from Twilio for client
+
+## Running NGROK
```
This section should probably be named something about **Twilio** since `ngrok` is a tool to enable testing of the Twilio integration.

ErrorsAndGlitches commented on environment-docker.yml:
```diff
@@ -2,57 +2,63 @@ name: waisn-tech-tools
     - urllib3==1.25.3
-prefix: /opt/conda/envs/waisn-tech-tools
+prefix: //anaconda3/envs/waisn-tech-tools
+
```
you can remove the **prefix** section

ErrorsAndGlitches commented on environment-docker.yml:
```diff
@@ -2,57 +2,63 @@ name: waisn-tech-tools
 channels:
```
I wrote a guide on how to update this file. See the **Docker** section:

* https://github.com/ErrorsAndGlitches/waisn-tech-tools/tree/2-sns-integration#updating-the-environment

There are still a lot more changes in this file than I would expect to see.

chandojo replied:
Working on this

ErrorsAndGlitches commented on waisntechtools/alerts/recent_twilio_messages.py:
```diff
@@ -0,0 +1,22 @@
+
+    if request_valid:
+        client = Client(settings.TWILIO_ACCOUNT_SID,
+                        settings.TWILIO_AUTH_TOKEN)
```
The client is being instantiated the same way multiple times, which means that it should be abstracted out.

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -1,8 +1,11 @@
 from alerts.waisn_auth import waisn_auth
+from alerts.recent_twilio_messages import recent_twilio_messages
+
+import requests
```
This is `import`-ed in a number of files but doesn't seem to be used.

ErrorsAndGlitches commented on waisntechtools/alerts/urls.py:
```diff
@@ -10,4 +10,6 @@
     path('alert/sent', alert_request.sent, name='alert_sent'),
     path('follow-up', follow_up_request.index, name='follow_up'),
     path('follow-up/sent', follow_up_request.sent, name='follow_up_sent'),
+    path('twilio-endpoint',
```
Could you keep the same end-point that was there before since the project has been deployed? This is a better name, which I think is worth changing, but something that I think makes sense to do by itself so that we can have a focused roll out of this change since it'll require configuration change in Twilio to happen simultaneously during deployment.

ErrorsAndGlitches commented on waisntechtools/alerts/recent_twilio_messages.py:
```diff
@@ -0,0 +1,22 @@
+    if settings.DEBUG == True:
+        request_valid = True
+
+    if request_valid:
```
It feels like there is a missing commit since there was a comment on this that was not addressed.

ErrorsAndGlitches commented on waisntechtools/waisntechtools/settings/common.py:
```diff
@@ -118,6 +121,10 @@ def environ_var(key):
     'profile'
 ]
 
+TWILIO_ACCOUNT_SID = os.environ['TWILIO_ACCOUNT_SID']
```
can you use the `environ_var` helper function? I believe this will actually break `python manage.py test`.

Commit:
Issue #4 Integrate with Twilio API for receiving SMS

[PROBLEM]
Users will send SMS to Twilio phone number which will be consumed by the service.

[SOLUTION]
* Messenger using Twilio client added
* Subscription App created to send correct reponse to subscriber

[TEST]
4 Unit test added for Message class

ErrorsAndGlitches commented on waisntechtools/alerts/get_twilio_cred.py:
```diff
@@ -0,0 +1,17 @@
+        mock = MagicMock()
+        mock.__iter__.return_value = []
+        return mock
+
```
nit - spacing

ErrorsAndGlitches commented on waisntechtools/alerts/get_twilio_cred.py:
```diff
@@ -0,0 +1,17 @@
+    else:
+        client = Client(settings.TWILIO_ACCOUNT_SID,
+                        settings.TWILIO_AUTH_TOKEN)
+        return client
```
inline

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,15 @@
+
+from django.test import TestCase
+
+from alerts.messenger import Messenger, Message
```
`Messenger` is imported but not used.

I REALLY recommend using PyCharm. PyCharm will complain about a number of syntactical issues as well as unused imports.

Also, PyCharm supports showing the diff locally in a really nice UI. It can help you review your code change before you push it up for review. I recommend this to all of the developers on my team.

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,15 @@
+from alerts.models import Subscriber
+
+
+class MessageTestCase(TestCase):
```
There are no tests for the `Messenger` class?

ErrorsAndGlitches commented on waisntechtools/alerts/messenger.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class Messenger(object):
+    def __init__(self):
+        self.client = get_twilio_cred()
```
while the `get_twilio_cred()` function does handle the test environment scenario, it'll probably be more object oriented to allow this to be passed in through the constructor. It can serve as a default value though:

```
def __init__(self, client=get_twilio_client()):
        self._client = client
```

Two other things:
1. the function says it gets credentials which is assigned to a client variable name. This is confusing.
2. make the member variable private i.e. `self._client`

ErrorsAndGlitches commented on waisntechtools/alerts/get_twilio_cred.py:
```diff
@@ -0,0 +1,17 @@
+from django.conf import settings
```
May want to call this file `twilio.py` since it is more generic and the name won't need to change if more twilio functions are added.

ErrorsAndGlitches commented on waisntechtools/alerts/tests/test_messenger.py:
```diff
@@ -0,0 +1,15 @@
+
+class MessageTestCase(TestCase):
+    def test_message(self):
+        temp = tempfile.NamedTemporaryFile()
```
it would be more pythonic to have use the resource manager syntax so the resource is cleaned up before the test ends.

```
with tempfile.NamedTemporaryFile() as temp:
        temp.write('The test has completed'.encode(encoding='utf-8'))
        temp.flush()
        read_message = Message(temp.name).contents()
        self.assertEqual(read_message, 'The test has completed')
```

ErrorsAndGlitches commented on waisntechtools/alerts/twilio_app.py:
```diff
@@ -0,0 +1,69 @@
+        courier.receive()
+        return HttpResponse(status=200)
+    except Exception as e:
+        print("There is an error: {}".format(e))
```
can you use Django logging here?

ErrorsAndGlitches commented on waisntechtools/alerts/urls.py:
```diff
@@ -1,6 +1,6 @@
-
-from alerts import views, alert_request, follow_up_request
+from alerts import views, alert_request, follow_up_request, twilio_app
+import requests
```
used?

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -25,10 +27,17 @@ def debug(request):
 
-    def get_queryset(self):
-        return Subscriber.objects.order_by('-date_registered')
+    def recent_twilio_messages(self):
```
make private

ErrorsAndGlitches commented on waisntechtools/alerts/views.py:
```diff
@@ -25,10 +27,17 @@ def debug(request):
+    def recent_twilio_messages(self):
+        client = get_twilio_cred()
+        messages_list = client.messages.list(limit=10)
+        return messages_list
```
inline
